spec/editors_draft.html

```diff
@@ -3298,6 +3298,231 @@ <h2>Evaluation Semantics</h2>
 
   </section>
 
+  <section id="import-export-composite-values">
+    <h1>Importing and Exporting of Composite Values</h1>
+
+    <section class="informative">
+      <h2>Motivation</h2>
+      <p>
+        In the normal handling of an RDF syntax, a blank node is scoped to the document, and any identifiers used in that document are merely a syntactic convenience for referencing blank nodes freshly allocated during parsing.
+        For example,
+      </p>
+        
+      <pre id="load-renaming-blank-example"
+         data-transform="updateExample"
+         data-content-type="text/turtle"
+         class="nohighlight example">
+      <!--
+      _:b1 ex:p4 "[ <http://example.org/alice>, <http://example.org/bob>, _:eve ]"^^cdt:List .
+      -->
+      </pre>
+      
+      <p>
+        Parsing this data as Turtle will result in a single triple whose subject is a blank node (now dissociated from the `b1` identifier), and whose object is a three-element <a>cdt:List literal</a>.
+        However, using a Turtle parser not aware of Composite Values will result in an object literal containing the substring "_:eve".
+        This presents a challenge, as allowing values to reference specific blank nodes by identifier in this way would allow multiple documents to reference the same blank node, and might expose <em>internal</em> blank node identifiers as <em>external</em> identifiers (a form of unintentional <a href="https://www.w3.org/TR/rdf11-concepts/#dfn-skolem-iri">skolemization</a>).
+      </p>
+        
+      <p>
+        The following sections describe requirements for conforming SPARQL systems that must be applied to composite values during their import into and export from datasets to support proper handling of blank nodes.
+      </p>
+    </section>
+
+    <section>
+      <h2>Importing Requirements</h2>
+      <p>
+        Conforming implementations MUST process <a>cdt:List literal</a>s and <a>cdt:Map literal</a>s during loading, replacing the substrings containing the serialized form of any blank nodes with identifier <var>id</var> with a blank node label <var>b</var>, such that
```

Comment:
Will try to update language about "substrings".

Agreed on using "identifier" consistently instead of "label."

Reply:
Updated text now reads:

> Conforming implementations MUST process cdt:List literals and cdt:Map literals during loading, replacing any substring id matching the BLANK_NODE_LABEL production of the grammar with a blank node identifier b, such that

Is that better?

Reply:
> Is that better?

Yes, much better! One more thing to make it even more clear: "... replacing, in their lexical form, any substring _id_ ..."

Unfortunately, there is yet another tricky bit here. The BLANK_NODE_LABEL production matches strings of the form `_:x` but these strings are not blank node identifiers; the blank node identifier is the substring after `_:` (e.g., `x` in `_:x`). So, strictly speaking, the substring _id_ must be replaced by a string "`_:`"+_b_ where _b_ is a blank node identifier.

Reply:
> Conforming implementations MUST process cdt:List literals and cdt:Map literals during loading, replacing, in their lexical form, any substring _:id matching the BLANK_NODE_LABEL production of the grammar with a string "_:"+b where b is a blank node identifier, such that

How's this?

Also, is it weird that we've defined the range of bnl2bn as taking a string *including* the leading "_:"? I'm not sure it's a *problem*, but it seems strange.

Reply:
> Also, is it weird that we've defined the range of bnl2bn as taking a string _including_ the leading "_:"? I'm not sure it's a _problem_, but it seems strange.

I don't think here is any problem with that.

Conceptually, there is no difference between defining bnl2bn over the set of blank node identifiers or over the set of strings that match the BLANK_NODE_LABEL production. The reason why I chose the latter is because it was easier to write the definition this way. Everything else would have required to extend the definition by talking about substrings of the strings that match BLANK_NODE_LABEL production.